File: C++/9012_괄호_stack.cpp
#include <iostream>
#include <stack>
#include <cstring>
using namespace std;

int main() {
	int T;
	cin >> T;
	for(int t = 0; t < T; t++){
		string s;
		cin >> s;
		
		bool flag = true; // flag가 true면 규칙에 위배되는 ')'는 존재하지 않음(규칙에 위배되는 '('는 존재할 수 있음)
		stack<char> bracket;
		int s_len = s.size();
		for(int i = 0; i < s_len; i++){
			if(s[i] == '('){ // 여는 괄호는 stack에 삽입
				bracket.push('(');
			}
			else{ // 닫는 괄호 만나면 여는 괄호 pop
				if(bracket.size() == 0){ // 만약 pop할 괄호가 없다면 닫는 괄호와 쌍이 될 여는 괄호 없다는 뜻이므로 flag를 false로 바꾸고 break
					flag = false;
					break;
				}
				bracket.pop(); // 아니라면 pop
			}
		}
		
		if(bracket.size() == 0 && flag == true){ // 규칙에 위배되는 ')'가 존재하지 않고(flag = true), 모든 '('의 쌍이 존재한다면(bracket.size() = 0)
			cout << "YES\n"; // YES 출력
		}
		else{ // 아니라면 NO 출력
			cout << "NO\n";
		}
	}
	return 0;
}

File: Python/9012_괄호_stack.py
if __name__ == '__main__':
	T = int(input())
	for _ in range(T):
		S = input()
		
		bracket = [] # 괄호를 담을 배열(stack으로 사용될 것임)
		for s in S:
			if s == '(': # '('라면 배열에 추가
				bracket.append('(')
			else: # ')'인데
				if len(bracket) == 0: # 짝이 되어줄 '('가 없다면, 배열에 ')'를 추가하고 break
					bracket.append(')')
					break
				bracket.pop() # 아니라면 배열에 있는 '(' pop
		
		if len(bracket) == 0: # bracket의 크기가 0이어야 모든 짝이 맞는 것
			print("YES")
		else: # 아닌 경우: 1) ')'와 짝이 될 '('가 없어서 '('가 하나 들어가 있는 상황, 2) '('가 남은 상황
			print("NO")
